### DISCUSSION TITLE
How to make a utility function like Vue Use

### DISCUSSION THREAD

**Discussion by Rolanddoda** [vueuse/vueuse#1065] [Q&A]
I stumbled on a weird issue. I am creating a simple library and I am returning a `ref` but that **is not reactive**.
For simplicity reasons, let's say I have a code like this:

```js
import { ref } from 'vue'

export const useCounter = () => {
  const counter = ref(0);

  const increaseCounter = () => {
    console.log('increaseCounter called')
    counter.value++;
  }

  return {
    counter,
    increaseCounter
  };
}
```

**Note that the code above is from a local library**.
Now in my Vue App, I try to use that function.

```html
<script>
import { defineComponent } from 'vue'
import { useCounter } from 'local-library'

export default defineComponent({
  name: 'App',
  setup() {
    const { counter, increaseCounter } = useCounter()

    return {
      counter,
      increaseCounter
    }
  }
})
</script>

<template>
  Counter: {{ counter }}
  <button @click="increaseCounter">Increase</button>
</template>
```

The above will NOT work! 🤯
I was surprised. After a lot of research, I found from [HERE](https://www.npmjs.com/package/@vue/reactivity) after reading it a lot of times, that my code above doesn't work because my app and library use `ref` but they use "different" Vue instance. 🤔

And I proved that by passing `ref` from my Vue app to the library and use that `ref` for declaring the reactive variable. And it worked!

```js
export const useCounter = (refFromVueApp) => {
  const counter = refFromVueApp(0);

  const increaseCounter = () => {
    console.log('increaseCounter called')
    counter.value++;
  }

  return {
    counter,
    increaseCounter
  };
}
```

But now things don't make sense! How is VueUse doing it ??
I tried `useMouse` from `@vueuse/core` and the `x` and `y` properties **ARE REACTIVE**.
I saw the [source code of useMouse](https://github.com/vueuse/vueuse/blob/main/packages/core/useMouse/index.ts) but I don't see how it is doing it except that it import `ref` from 'vue-demi'. I checked `vue-demi` as well but I don't see anything related to different vue instances.

Does anyone knows what is wrong ?

**Comment by sibbng** [comment]
You need to mark `vue` package as external package on your local package.

Vueuse does it like that: https://github.com/vueuse/vueuse/blob/main/scripts/rollup.config.ts#L27-L30

I don't know which bundler you are using, but every bundler has this feature.

**Reply by Rolanddoda** [reply]
@sibbng thanks for the answer. I am not using any bundler. Regarding "external" package, I tried "peerDependencies" in `package.json` but that doesn't work either.

**Reply by sibbng** [reply]
If your app running on top of Vite including this package in optimizeDeps.include might help. 

If it doesn't work try bundling your package with tsup, it's automatically externalize your dependencies: https://tsup.egoist.sh/#excluding-packages Dont forget to define output targets on your package.json. e.g.:

```json
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    }
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  ```

**Reply by Rolanddoda** [reply]
Unfortunately, my App is running on webpack.

I was able to make it work but in a strange way. I had `vue` as a peer dependency on the library wanting to say that vue is installed on the app environment, but that didn't work somehow.
What I did this morning is to just remove `vue` from package.json and everything worked. Even though it's not the solution I want because TS is yelling about module not found.

So the solution would be to use a bundler right ?

**Reply by Rolanddoda** [reply]
@sibbng I installed `tsup` and it didn't work. I even copied `package.json` from `vue-use`. Here is my `package.json`:

```json
{
  "name": "my-lib",
  "version": "0.1.0",
  "description": "Playing around",
  "license": "MIT",
  "main": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist/**/*"
  ],
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "import": "./dist/index.mjs",
      "types": "./dist/index.d.ts"
    }
  },
  "author": "Roland Doda <rolanddoda2014@gmail.com>",
  "scripts": {
    "lint": "eslint --fix --ext .js,ts src/**",
    "release": "np --no-tests --no-yarn",
    "build": "tsup src/index.ts --format cjs,esm --dts --clean --external vue",
    "dev": "npm run build -- --watch src"
  },
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^5.0.0",
    "@typescript-eslint/parser": "^5.0.0",
    "eslint": "^8.0.1",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^4.0.0",
    "prettier": "^2.4.1",
    "tsup": "^5.11.6",
    "typescript": "^4.4.4"
  },

  "peerDependencies": {
    "@vue/composition-api": "^1.1.0",
    "vue": "^2.6.0 || ^3.2.0"
  },
  "peerDependenciesMeta": {
    "vue": {
      "optional": true
    },
    "@vue/composition-api": {
      "optional": true
    }
  },
  "dependencies": {
    "vue-demi": "*"
  }
}

```

And this is my `src/index.ts` file:

```ts
import { ref } from 'vue-demi'

export const useCounter = () => {
  const counter = ref(5)

  const increaseCounter = () => {
    console.log('increaseCounter called')
    counter.value++
  }

  return {
    counter,
    increaseCounter
  }
}
```

But no, reactivity outside of my **npm local package** doesn't work. Inside the package though reactivity works fine.

**Reply by sibbng** [reply]
I can't help without a preproduction from this point. Could be something related to your frontend app.

**Comment by Rolanddoda** [comment]
UPDATE: Okii after 2 days of debugging... here is what happened. I ran `npm unlink` in the "library", removed `node_modules`, `package.json` from the 3 different projects "app with webpack", "app with vite", and the "library"!

I thought of just taking everything from scratch and see what is going on in details. So I `npm install`-ed dependencies again and run `npm link` in the library and `npm link package-name` in the apps. When I went to debug and run the apps... suddenly... everything was working. 🤯🤯

I do have a strong suspicion that the "issue" is related to `npm link`.

Is there any other/better way to test local packages ?
